servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/JavaNetSoTimeoutHttpConnectionFilter.java

```diff
@@ -0,0 +1,213 @@
+ * limitations under the License.
+ */
+package io.servicetalk.http.utils;
+
+import io.servicetalk.concurrent.Cancellable;
+import io.servicetalk.concurrent.CompletableSource;
+import io.servicetalk.concurrent.TimeSource;
+import io.servicetalk.concurrent.api.Executor;
+import io.servicetalk.concurrent.api.Processors;
+import io.servicetalk.concurrent.api.Publisher;
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.concurrent.api.SourceAdapters;
+import io.servicetalk.concurrent.internal.ThrowableUtils;
+import io.servicetalk.http.api.FilterableStreamingHttpConnection;
+import io.servicetalk.http.api.HttpContextKeys;
+import io.servicetalk.http.api.HttpExecutionStrategies;
+import io.servicetalk.http.api.HttpExecutionStrategy;
+import io.servicetalk.http.api.HttpHeaderNames;
+import io.servicetalk.http.api.HttpHeaderValues;
+import io.servicetalk.http.api.HttpRequestMetaData;
+import io.servicetalk.http.api.StreamingHttpConnectionFilter;
+import io.servicetalk.http.api.StreamingHttpConnectionFilterFactory;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpResponse;
+import io.servicetalk.http.utils.AbstractTimeoutHttpFilter.FixedDuration;
+import io.servicetalk.transport.api.ExecutionContext;
+
+import java.net.SocketOptions;
+import java.net.SocketTimeoutException;
+import java.time.Duration;
+import java.util.concurrent.TimeoutException;
+import java.util.function.BiFunction;
+import javax.annotation.Nullable;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A filter to mimics {@link SocketOptions#SO_TIMEOUT} behavior on the client-side.
+ * <p>
+ * While {@link TimeoutHttpRequesterFilter} apples a timeout for the overall duration to receive either the response
+ * metadata (headers) or the complete reception of the response (including headers, payload body, optional trailers, as
+ * well as time to send the request), this filter applies timeout only to read operations. It means that time to send
+ * the request is not accounted. Also, if the remote server is sending a large payload body, the timeout will be applied
+ * on every chunk read, which may result in unpredictable time to read the full response if the remote slowly sends
+ * 1 byte withing the timeout boundaries. Use this filter only for compatibility with classic blocking Java libraries.
+ * To protect from the described use-cases, consider also using {@link TimeoutHttpRequesterFilter} before applying this
+ * filter.
+ * <p>
+ * This filter implements only {@link StreamingHttpConnectionFilterFactory} and therefore can be applied only at the
+ * connection level. This restriction ensures that the timeout is applied only for the response read operations
+ * (similar to {@link SocketOptions#SO_TIMEOUT} used by Java blocking API), without waiting for selecting or
+ * establishing a connection.
+ * <p>
+ * {@link SocketTimeoutException} (or its subtype) will be propagated when the timeout is reached.
+ *
```

Comment:
It would be great if this javadoc could also elaborate more on the http 100 continue behavior, it took me a while to figure this bit out in the code and users could benefit from a (terse) explanation as well.